kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/HighOffsetMetricsExposer.scala

```diff
@@ -8,80 +8,117 @@ package com.mwam.kafkakewl.metrics.services
       changes: KafkaTopicPartitionInfoChanges,
-      config: MetricExposerConfig
-  ): (MetricMap, List[(Metric.Counter[Long], Long)]) = {
-    val addedOrUpdated = changes.addedOrUpdated
+      config: TopicsConfig
+  ): (MetricMap, List[MetricUpdate]) = {
+    val addedOrUpdated = changes.addedOrUpdated.groupBy(_._1.topic)
     val removed = changes.removed
     var newMetricMap = metricMap
-    var metricUpdates: List[(Metric.Counter[Long], Long)] = List.empty
+    var metricUpdates: List[MetricUpdate] = List.empty
+
     // Add new metrics and update stale ones
-    for ((topicPartition, partitionInfo) <- addedOrUpdated) {
-
-      newMetricMap = newMetricMap.updatedWith(topicPartition.topic) {
-        case Some((topicMap: Map[Int, (Metric.Counter[Long], Long)], aggregateCounter: Metric.Counter[Long])) =>
-          val newTopicMap = topicMap.updatedWith(topicPartition.partition) {
-            case Some((counter, lastValue)) =>
-              val increase = partitionInfo.endOffset - lastValue
-              metricUpdates = metricUpdates :+ (aggregateCounter, increase)
-              if (config.perPartitionMetrics) {
-                metricUpdates = metricUpdates :+ (counter, increase)
-              }
-              Some((counter, partitionInfo.endOffset))
-            case None =>
-              val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
-              metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
-              if (config.perPartitionMetrics) {
-                metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
-              }
-              Some((counter, partitionInfo.endOffset))
-          }
-          Some((newTopicMap, aggregateCounter))
-        case None =>
-          val counter = Helper.newPartitionMetric(topicPartition.topic, topicPartition.partition)
-          val newTopicMap = Map.newBuilder
-            .addOne(
-              topicPartition.partition,
-              (counter, partitionInfo.endOffset)
-            )
-            .result()
-          val aggregateCounter = Helper.newTopicMetric(topicPartition.topic)
-          metricUpdates = metricUpdates :+ (aggregateCounter, partitionInfo.endOffset)
-          if (config.perPartitionMetrics) {
-            metricUpdates = metricUpdates :+ (counter, partitionInfo.endOffset)
-          }
-          Some((newTopicMap, aggregateCounter))
+    for ((topic, topicPartitionChanges) <- addedOrUpdated) {
+      newMetricMap = newMetricMap.updatedWith(topic) { partitionMetricsAndCounter =>
+        val (oldPartitionMetrics, aggregateCounter) =
+          partitionMetricsAndCounter.getOrElse((PartitionMetrics.empty, config.newTopicHighOffsetMetric(topic)))
+
+        val (newPartitionMetrics, partitionMetricUpdates, totalIncrease) = oldPartitionMetrics.updateWithNewOffset(topicPartitionChanges, config)
+        metricUpdates = partitionMetricUpdates ++ metricUpdates
```

Comment:
I think this is still a O(n) list concat operation. How about using `++:` (alias for `prependAll`)?